server/mailrepository/deleted-messages-vault-repository/src/main/java/org/apache/james/vault/MailAdapter.java

```diff
@@ -0,0 +1,162 @@
+
+    private final MailboxId.Factory mailboxIdFactory;
+    private final MessageId.Factory messageIdFactory;
+
+    public MailAdapter(MailboxId.Factory mailboxIdFactory, MessageId.Factory messageIdFactory) {
+        this.mailboxIdFactory = mailboxIdFactory;
+        this.messageIdFactory = messageIdFactory;
+    }
+
+    Mail toMail(DeletedMessage deletedMessage) throws MessagingException {
+        MimeMessageCopyOnWriteProxy mimeMessage = new MimeMessageCopyOnWriteProxy(new MimeMessageInputStreamSource(deletedMessage.getMessageId().serialize(), deletedMessage.getContent().get()));
+
+        Collection<AttributeValue<?>> serializedMailboxIds = deletedMessage.getOriginMailboxes().stream()
+            .map(MailboxId::serialize)
+            .map(AttributeValue::of)
+            .collect(Guavate.toImmutableList());
+
+        AttributeValue<?> subjectAttributeValue = AttributeValue.of(deletedMessage.getSubject().map(AttributeValue::of));
+
+        return MailImpl.builder()
+            .name(deletedMessage.getMessageId().serialize())
+            .sender(deletedMessage.getSender())
+            .recipients(deletedMessage.getRecipients())
+            .mimeMessage(mimeMessage)
+            .attribute(new Attribute(OWNER_ATTRIBUTE_NAME, AttributeValue.of(new SerializableUser(deletedMessage.getOwner()))))
+            .attribute(new Attribute(HAS_ATTACHMENT_ATTRIBUTE_NAME, AttributeValue.of(deletedMessage.hasAttachment())))
+            .attribute(new Attribute(DELIVERY_DATE_ATTRIBUTE_NAME, AttributeValue.of(new SerializableDate(deletedMessage.getDeliveryDate()))))
+            .attribute(new Attribute(DELETION_DATE_ATTRIBUTE_NAME, AttributeValue.of(new SerializableDate(deletedMessage.getDeletionDate()))))
+            .attribute(new Attribute(ORIGIN_MAILBOXES_ATTRIBUTE_NAME, AttributeValue.of(serializedMailboxIds)))
+            .attribute(new Attribute(SUBJECT_ATTRIBUTE_VALUE, subjectAttributeValue))
+            .build();
+    }
+
+    DeletedMessage fromMail(Mail mail) {
+        MemoizedSupplier<byte[]> byteContent = new MemoizedSupplier<>(Throwing.supplier(
+            () -> MimeMessageUtil.asBytes(mail.getMessage())));
+
+        return DeletedMessage.builder()
+            .messageId(messageIdFactory.fromString(mail.getName()))
+            .originMailboxes(retrieveMailboxIds(mail))
+            .user(retrieveOwner(mail))
+            .deliveryDate(retrieveDate(mail, DELIVERY_DATE_ATTRIBUTE_NAME))
+            .deletionDate(retrieveDate(mail, DELETION_DATE_ATTRIBUTE_NAME))
+            .sender(mail.getMaybeSender())
+            .recipients(ImmutableList.copyOf(mail.getRecipients()))
+            .content(() -> new ByteArrayInputStream(byteContent.get()))
+            .hasAttachment(retrieveHasAttachment(mail))
+            .subject(retrieveSubject(mail))
+            .build();
+
+    }
+
+    private Optional<String> retrieveSubject(Mail mail) {
+        return mail.getAttribute(SUBJECT_ATTRIBUTE_VALUE)
```

Comment:
Attribute should provide a `<T> Optional<T> getAttributeValueAs(AttributeName name, Class<T> type)` method to avoid this kind of ugly code (that will happen over and over again). WDYT?

Reply:
I would 100% upvote such a proposition (that I already made 6 months ago BTW)

Reply:
I did contribute this, but as a method of AttributeValue